Site Migration: My Sites should reset failed migration state

bisko opened the pull request:
#### Changes proposed in this Pull Request

* This PR makes the My Sites button reset the migration state to `inactive` if there's a failed migration for the current site.
* This will unlock the site, similar to the Try Again button

#### Testing instructions

* Apply diff or use the Calypso.live link
* Start a migration
* Cause it to fail in some way (to be added)
* Click My Sites
* You should be back to unlocked state

bisko committed:
Site Migration: My Sites resets a failed migration state

bisko committed:
Properly reset the redux state

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -191,7 +212,8 @@ export default connect(
 		// by the user yet
-		const siteId = getSelectedSiteId( state ) || getPrimarySiteId( state );
+		const currentSelectedSiteId = getSelectedSiteId( state );
+		const siteId = currentSelectedSiteId || getPrimarySiteId( state );
```
There's a selector for that FWIW: `getSelectedOrPrimarySiteId`

bisko replied:
We specifically want to reset only the currently selected site, as trying to reset the primary one might not be the intention of the customer where the primary site and the current site are different.

That's why I extracted the currently selected `siteId` as a separate property.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,23 @@ class MasterbarLoggedIn extends React.Component {
+			const { migrationStatus, currentSelectedSiteId } = this.props;
+
+			if ( currentSelectedSiteId && migrationStatus === 'error' ) {
+				wpcom.undocumented().resetMigration( currentSelectedSiteId );
```
Does this need any sort of error handling? Also, such a shame to have to import the entire `lib/wpcom` library just for this.  We're trying to reduce the places where we're doing that because it's quite big. And the logged in masterbar is displayed literally on every Calypso page. Could we define this as a data layer handler? Or perhaps `requestHttpData()`?

bisko replied:
Thanks for the suggestion! Updated the code and now it should be lighter :) Not sure how to trigger the matticbot check again though :(

tyxla replied:
It would usually retrigger at the end of the ICFY job. You can see that it automatically updates the matticbot message.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -33,6 +33,10 @@ import { getStatsPathForTab } from 'lib/route';
 import { domainManagementList } from 'my-sites/domains/paths';
 import WordPressWordmark from 'components/wordpress-wordmark';
 
```
```suggestion
```

bisko committed:
Update client/layout/masterbar/logged-in.jsx

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

bisko committed:
Update code to use `requestHttpData` instead of `wpcom.undocumented`

bisko committed:
Merge branch 'fix/site-migration-my-sites-reset-failed-migration' of github.com:Automattic/wp-calypso into fix/site-migration-my-sites-reset-failed-migration

tyxla reviewed (commented):
Code is looking great 😉 I'm only nitpicking with some suggestions, feel free to take them or throw them away. 

Also, haven't tested, so you might want to get testing approval.

mattsherman commented:
@bisko When testing this, clicking on My Sites did reset the migration. It did *not* unlock the Calypso admin or remove the "Import Failed" message. This behaves differently than the "Try Again" button on the "Import Failed" screen, which unlocks the Calypso admin.

bisko committed:
Properly reset local state when resetting server-side state

bisko committed:
Update code according to feedback

bisko commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,35 @@ class MasterbarLoggedIn extends React.Component {
+			const { migrationStatus, currentSelectedSiteId } = this.props;
+
+			if ( currentSelectedSiteId && migrationStatus === 'error' ) {
+				requestHttpData(
```
For the moment this should be the only place it's used. We can probably move it to `data-getters`, but we're going to see how it would play in refactoring the other places where we call the `reset-migration` endpoint and handle this in a separate PR.

bisko commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,35 @@ class MasterbarLoggedIn extends React.Component {
+						onSuccess: () => {},
+					} ),
+					{
+						fromApi: () => {},
```
Nope, removed. See answer about your `body: {}` comment below.

bisko commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,35 @@ class MasterbarLoggedIn extends React.Component {
+					} ),
+					{
+						fromApi: () => {},
+						freshness: -Infinity,
```
It's an artifact of copy-paste-driven-development. I got it from another place where this was used and it was explained that it helps to make sure the request is not cached. Updated to be a `0` as it should be having the same behavior.

bisko commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,35 @@ class MasterbarLoggedIn extends React.Component {
+						method: 'POST',
+						path: `/sites/${ currentSelectedSiteId }/reset-migration`,
+						body: {},
+						onSuccess: () => {},
```
Not :) Updated.

bisko commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -47,6 +51,35 @@ class MasterbarLoggedIn extends React.Component {
+						apiNamespace: 'wpcom/v2',
+						method: 'POST',
+						path: `/sites/${ currentSelectedSiteId }/reset-migration`,
+						body: {},
```
At the moment if `body` is omitted, the call is made to `rest/1.1` namespace, not `wpcom/v2` API.

I'm working on figuring out why this happens and I'm going to submit a PR to fix that separately, as it's another issue.

bisko commented:
> @bisko When testing this, clicking on My Sites did reset the migration. It did not unlock the Calypso admin or remove the "Import Failed" message. This behaves differently than the "Try Again" button on the "Import Failed" screen, which unlocks the Calypso admin.

"Expected" behavior. Accidentally I removed the code that resets the local Calypso state with the previous refactor to `requestHttpData`. I added it again and it should be behaving properly now.

andfinally commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,8 +241,10 @@ export default connect(
 			user: getCurrentUser( state ),
 			isSupportSession: isSupportSession( state ),
-			isMigrationInProgress: !! isSiteMigrationInProgress( state, siteId ),
+			isMigrationInProgress: !! isSiteMigrationInProgress( state, currentSelectedSiteId ),
```
I'm not sure why we're doing the `!!` here, as `isSiteMigrationInProgress` already returns a boolean – we're doing it in a couple of other places. Not a killer I guess.

bisko replied:
I think this is a leftover from a previous iteration on the code and looking at other places in Calypso, it's generally to ensure the return value is `bool`. 
Seems this is more of a code-style and making sure the value is in the preferred data type.

We can always change it if there's desire to unify code styles in the future.

andfinally replied:
Cool with me!

andfinally reviewed (commented):
This LGTM, and tests as expected. Minor observation about the `!!` in one prop, which is not necessarily a blocker. There may be some reason for it I'm not understanding.
